homeassistant/components/octoprint/__init__.py

```diff
@@ -202,3 +232,20 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+def async_get_client_for_service_call(
+    hass: HomeAssistant, call: ServiceCall
+) -> OctoprintClient:
+    """Get the client related to a service call (by device ID)."""
+    device_id = call.data[CONF_DEVICE_ID]
+    device_registry = dr.async_get(hass)
+
+    if device_entry := device_registry.async_get(device_id):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if entry.entry_id in device_entry.config_entries:
+                return cast(
+                    OctoprintClient, hass.data[DOMAIN][entry.entry_id]["client"]
+                )
```

Comment:
```suggestion
        for entry_id in device_entry.config_entries:
            if data:= hass.data[DOMAIN].get(entry_id):
        	return cast(OctoprintClient, data["client"])
```

As there are fewer config_entries for a device, this should be quicker.

Reply:
Done